webpack/webpack-embedded-dependencies-plugin/src/index.ts

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+import path from 'path';
+
+import type { IPackageJson } from '@rushstack/node-core-library';
+
+import { LegacyAdapters, FileSystem } from '@rushstack/node-core-library';
+import { Compilation, Compiler, WebpackPluginInstance, sources, WebpackError } from 'webpack';
+
+import { LICENSE_FILES_REGEXP, COPYRIGHT_REGEX } from './regexpUtils';
+
+const PLUGIN_NAME: 'EmbeddedDependenciesWebpackPlugin' = 'EmbeddedDependenciesWebpackPlugin';
+
+interface IEmbeddedDependenciesFile {
+  name?: string;
+  version?: string;
+  embeddedDependencies: IPackageData[];
+}
+
+interface IResourceResolveData {
+  descriptionFileData?: IPackageData;
+  descriptionFileRoot?: string;
+  relativePath?: string;
+}
+
+interface IWebpackModuleCreateData {
+  resourceResolveData?: IResourceResolveData;
+}
+
+interface IPackageData extends IPackageJson {
+  copyright: string | undefined;
+  author?: string | { name?: string };
+  licenses?: { type: string; url: string }[];
+  licensePath?: string;
+}
+
+interface IEmbeddedDependenciesWebpackPluginOptions {
+  outputFileName?: string;
+  generateLicenseFile?: boolean;
+  generateLicenseFileFunction?: LicenseFileGeneratorFunction;
+  generatedLicenseFilename?: LicenseFileName;
+}
+
+type LicenseFileGeneratorFunction = (packages: IPackageJson[]) => string;
+type PackageMapKey = `${string}@${string}`;
+type LicenseFileName = `${string}.${'html' | 'md' | 'txt'}`;
+type ThirdPartyPackageMap = Map<PackageMapKey, { dir: string; data: IPackageData }>;
+type FlattenedPackageEntry = [PackageMapKey, { dir: string; data: IPackageData }];
+
+/**
+ * @alpha
+ * Webpack plugin that generates a file with the list of embedded dependencies
+ * and their licenses.
+ */
+export default class EmbeddedDependenciesWebpackPlugin implements WebpackPluginInstance {
+  public outputFileName: string;
+  public generateLicenseFile: boolean;
+  public generateLicenseFileFunction?: LicenseFileGeneratorFunction;
+  public generatedLicenseFilename: LicenseFileName;
+
+  public constructor(options?: IEmbeddedDependenciesWebpackPluginOptions) {
+    this.outputFileName = options?.outputFileName || 'embedded-dependencies.json';
+    this.generateLicenseFile = options?.generateLicenseFile || false;
+    this.generateLicenseFileFunction = options?.generateLicenseFileFunction || undefined;
+    this.generatedLicenseFilename = options?.generatedLicenseFilename || 'THIRD-PARTY-NOTICES.html';
+  }
+
+  public apply(compiler: Compiler): void {
+    const thirdPartyPackages: ThirdPartyPackageMap = new Map();
```

Comment:
You can avoid having to side-channel this data between the compilation hook and the `normalModuleFactory` hook by accessing `normalModuleFactory` via the second argument on the `compilation` hook.